opt: detect and extract projected tuple equalities

RoryBlevins opened the pull request:
This commit adds a new rule to join normalisation that detects joins on tuple equalities between a tuple and a variable that is a projected tuple and splits them out into multiple equalities on the underlying columns, and hoists the project above the join. This allows us to simplify a number of queries where previously the projected tuple was preventing us merging projections.

Resolves: #43198

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=46186) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=46186) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=46186) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46186)
<!-- Reviewable:end -->

RoryBlevins commented:
@RaduBerinde PR for further review - be gentle, it's my first real Go

irfansharif commented:
Here's the failure (I think the others are known flakes):

<img width="1158" alt="image" src="https://user-images.githubusercontent.com/10536690/76861996-ce2e8f00-6833-11ea-84fa-2de76a59731f.png">

We're in our code-stability month in lieu of our upcoming release, so the PRs we're accepting in the interim must have a "Release justification: <justification>" note in the PR/commit.

RoryBlevins commented:
Thanks @irfansharif  - is there some documentation on what should go in the release justification note, such as a list of possible justifications?

rohany commented:
> Thanks @irfansharif - is there some documentation on what should go in the release justification note, such as a list of possible justifications?

The commit message template lists these!

RaduBerinde commented:
No need for a release justification, this doesn't need to go in 20.1. We will merge the PR after the branching happens in a couple of weeks.

RaduBerinde reviewed (approved):
Thank you for you contribution, this is great overall!

There are a bunch of other lint failures, you can run `make lint` locally (or at least `make lintshort`) to see them. Also be sure to run all the tests under `sql/opt/...`, there are some failures thereth as well. You ca use the `-rewrite` flag in most cases to rewrite the tests where plans have changed. The `TestStatsQuality` might be harder to fix, but we can help with that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-:-M2d54nv-_tKzaWvzDAQ:b-9z6x8a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde, @RoryBlevins, and @rytaft)

___
*[pkg/sql/opt/norm/join.go, line 802 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d3z5PDMOSWYGUkE8I:-M2d3z5PDMOSWYGUkE8J:bfyv8f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L802)):*
> ```Go
> }
> 
> /// CanExtractTupleEquality Checks if the given filters contain one or more
> ```

[nits] `//` instead of `///`
`checks` instead of `Checks`
end comment with `.`
___
*[pkg/sql/opt/norm/join.go, line 818 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4Fh7DOjhgOLXDbdn:-M2d4Fh7DOjhgOLXDbdo:b-7pcu7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L818)):*
> ```Go
> 		}
> 	}
> 	hasProjectTupleEquality := false
> ```

Do we need this? Didn't the rule matching already establish this? (if yes, we should also adjust the function comment)
___
*[pkg/sql/opt/norm/join.go, line 820 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4AF1E5lCEbIZR5QT:-M2d4AF1E5lCEbIZR5QU:b-7i9fp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L820)):*
> ```Go
> 	hasProjectTupleEquality := false
> 	for _, f := range filters {
> 		//we're only interested in filters which reference a projected column
> ```

[nit] space after `//`
___
*[pkg/sql/opt/norm/join.go, line 821 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4V6V9v_0TDd9q07F:-M2d4V6W2MDL1fjjWBpn:b-34fgl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L821)):*
> ```Go
> 	for _, f := range filters {
> 		//we're only interested in filters which reference a projected column
> 		if !f.ScalarProps().OuterCols.Intersection(projectedColsSet).Empty() {
> ```

can just use `Intersects()`
___
*[pkg/sql/opt/norm/join.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4e8H0BvhkACgfWjN:-M2d4e8H0BvhkACgfWjO:b-a6ufda) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L832)):*
> ```Go
> }
> 
> func isValidTupleEqualityExpression(f *memo.FiltersItem,
> ```

[nit] a short comment would be good
[nit] we have some linters that enforce a certain alignment here (and will prevent the change for merging). You can also set up `crlfmt` (http://github.com/cockroachdb/crlfmt) to do it automatically.
___
*[pkg/sql/opt/norm/join.go, line 848 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d5MVG6sZbEpVKkB6P:-M2d5MVG6sZbEpVKkB6Q:b24iqto) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L848)):*
> ```Go
> }
> 
> /// RewriteTupleEquality checks a list of filters for equalities between a projected
> ```

`//`, `takes`, end with `.`
___
*[pkg/sql/opt/norm/join.go, line 866 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d5JChAc5aer6r16gz:-M2d5JChAc5aer6r16h-:b-y98bl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L866)):*
> ```Go
> 
> 	for _, f := range filters {
> ```

[nit] extra blank line
___
*[pkg/sql/opt/norm/join.go, line 888 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d5HKp5G6oaG0yaL5n:-M2d5HKp5G6oaG0yaL5o:bxim66m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L888)):*
> ```Go
> 				}
> 			} else {
> 				panic(errors.AssertionFailedf("cannot extract tuple equality from non-equality comparision"))
> ```

comparison
___
*[pkg/sql/opt/norm/rules/join.opt, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d5d2UAfYBS-mttk-R:-M2d5d2UAfYBS-mttk-S:b-q0tos9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L448)):*
> ```
> # TODO(andyk): Add other join types.
> [HoistJoinProjectRight, Normalize]
> (InnerJoin | InnerJoinApply | LeftJoin | LeftJoinApply | SemiJoin
> ```

This doesn't look right, and I'm surprised it didn't cause some failures. A SemiJoin does not return the right-hand side columns, i.e. it only produces `(OutputCols $right)`; we cannot then project `(OutputCols2 $left $right)`

We can have a separate rule that just drops the passthrough projection from the right side of a semi-join
___
*[pkg/sql/opt/norm/rules/join.opt, line 469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6JzGBTL45DtxznSt:-M2d6JzGBTL45DtxznSu:b9jblp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L469)):*
> ```
> # input of the join.
> [HoistJoinProjectLeft, Normalize]
> (InnerJoin | InnerJoinApply | LeftJoin | LeftJoinApply | SemiJoin
> ```

Same here, this rule would make sense for `SemiJoin` but not with `(OutputCols2 $left $right)`.
___
*[pkg/sql/opt/norm/rules/join.opt, line 536 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6V2p7UbUCEY5FHNr:-M2d6V2p7UbUCEY5FHNs:b-qu24fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L536)):*
> ```
> 
> # ExtractProjectedTupleEqualityLeft detects joins which contain equalities between 
> # projected tuple variables and tuples, and splits these tuples out into multiple 
> ```

[nit] remove `and` (there's another `and` following).
___
*[pkg/sql/opt/norm/rules/join.opt, line 541 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6QfC2xBHIvDb9E8x:-M2d6QfC2xBHIvDb9E8y:bxbhi2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L541)):*
> ```
> [ExtractProjectedTupleEqualityLeft, Normalize]
> (InnerJoin | LeftJoin | SemiJoin
>     $left: (Project  $input:* $projections:* $passthrough:*)
> ```

[nit] extra space after Project
___
*[pkg/sql/opt/norm/rules/join.opt, line 546 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6ZCZ907x_z1TsYa0:-M2d6ZCZ907x_z1TsYa1:b-kc5h8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L546)):*
> ```
>         ...
>             (FiltersItem (Eq (Variable) (Tuple))) &
>             (CanExtractTupleEquality
> ```

[nit] `$projections $on)` can go on this line
___
*[pkg/sql/opt/norm/rules/join.opt, line 566 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6jLS9aa3tbXkQiQQ:-M2d6jLS9aa3tbXkQiQR:bquke38) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L566)):*
> ```
> )
> 
> # ExtractProjectedTupleEqualityRight like ExtractProjectedTupleEqualityLeft, but for the right
> ```

[nit] `is like`, `for the right side of the join.`
___
*[pkg/sql/opt/norm/testdata/rules/join, line 2914 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6uVRC3ZQVIySkHfe:-M2d6uVRC3ZQVIySkHff:brg031q) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/testdata/rules/join#L2914)):*
> ```
>  └── filters
>       └── (substring('', ')') = '') = (u:3 > 0) [outer=(3)]
> ```

Please look inside this file and follow the same format. Each rule gets its own `------` block.
You should also use `expect` or`expect-not` for the particular rule we're testing.

We should come up with tests that exercise just the left rule, and just the right rule, in addition to this. We also need tests for cases where the rule satisfies a subset of the requirements but does not fire (e.g. if a projected tuple column is used in multiple filters)
___
*[pkg/sql/opt/norm/testdata/rules/join, line 2932 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d74D2-HHJQsHP2vhH:-M2d74D2-HHJQsHP2vhI:b-nniodu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/testdata/rules/join#L2932)):*
> ```
> 
> exec-ddl
> ALTER TABLE cd INJECT STATISTICS '[
> ```

we don't need this here, we can see how the rule gets applied regardless if we choose a lookup join or not. this could be a separate test in `opt/exec/execbuilder/testdata/lookup_join`
___
*[pkg/sql/opt/norm/testdata/rules/join, line 2943 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d7MuM1O-kBWBeMBMJ:-M2d7MuN2a7Qh7P5UrDA:b-52roqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/testdata/rules/join#L2943)):*
> ```
> 
> opt
> SELECT a,b FROM ab, cd WHERE a=c AND b=d
> ```

not needed here
___
*[pkg/sql/opt/norm/testdata/rules/join, line 2956 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d7Pwp8LoWTgMQNgV9:-M2d7Pwp8LoWTgMQNgVA:br38wfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/testdata/rules/join#L2956)):*
> ```
>       └── filters (true)
> 
> opt
> ```

use `norm` not `opt`; add `expect`


<!-- Sent from Reviewable.io -->

RoryBlevins reviewed (commented):
Huh - I thought I had already run `make lint` over the project

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-:-M2e9T5w3KwEi5_Opx2V:br8qto7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde, @RoryBlevins, and @rytaft)

___
*[pkg/sql/opt/norm/join.go, line 818 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4Fh7DOjhgOLXDbdn:-M2e2Ywo22KmHaW06Bub:b1w9faf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L818)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we need this? Didn't the rule matching already establish this? (if yes, we should also adjust the function comment)
</blockquote></details>

Yeah, I think that's right - was a leftover from before we had the additional matching
___
*[pkg/sql/opt/norm/join.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4e8H0BvhkACgfWjN:-M2e6Tm11Z1ZpykQTQfA:brv9s95) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L832)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] a short comment would be good
[nit] we have some linters that enforce a certain alignment here (and will prevent the change for merging). You can also set up `crlfmt` (http://github.com/cockroachdb/crlfmt) to do it automatically.
</blockquote></details>

I thought I had already run gofmt - but apparently not. Is there any documentation on how to run `crlfmt`?  I'm new to go and may be missing a convention, but the documentation seems to be a single line.
___
*[pkg/sql/opt/norm/rules/join.opt, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d5d2UAfYBS-mttk-R:-M2e7_yP2yGG3R5h234p:b-xvesws) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L448)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This doesn't look right, and I'm surprised it didn't cause some failures. A SemiJoin does not return the right-hand side columns, i.e. it only produces `(OutputCols $right)`; we cannot then project `(OutputCols2 $left $right)`

We can have a separate rule that just drops the passthrough projection from the right side of a semi-join
</blockquote></details>

My mistake - leftover from some early experimentation
___
*[pkg/sql/opt/norm/rules/join.opt, line 469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6JzGBTL45DtxznSt:-M2e7e0Q8N02r-GQH70G:byr4ffv) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/rules/join.opt#L469)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Same here, this rule would make sense for `SemiJoin` but not with `(OutputCols2 $left $right)`.
</blockquote></details>

see above


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Thanks for your contribution! Added some comments below.

> Reviewed 1 of 4 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-:-M36aUiA79HYZKijv_80:bdixi4w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @RoryBlevins)

___
*[pkg/sql/opt/norm/join.go, line 802 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36RwRI8qAGM7DD4tSN:-M36RwRI8qAGM7DD4tSO:b7jk6rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L802)):*
> ```Go
> }
> 
> // CanExtractTupleEquality checks that
> ```

[nit] comment should be wrapped at 80 characters.
___
*[pkg/sql/opt/norm/join.go, line 803 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36SATgDehMumxhetWX:-M36SATgDehMumxhetWY:b-pgnzz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L803)):*
> ```Go
> 
> // CanExtractTupleEquality checks that
> // no other filters reference any of the projected columns.
> ```

This comment doesn't seem to match what's actually happening in this function.
___
*[pkg/sql/opt/norm/join.go, line 813 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36U8Qo30ZdY3SJyI2h:-M36U8Qo30ZdY3SJyI2i:brb5ugt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L813)):*
> ```Go
> 		projectedColsSet.Add(p.Col)
> 		if _, isTuple := p.Element.(*memo.TupleExpr); isTuple {
> 			projectedTupleCols[p.Col] = p
> ```

Since you're never actually using `p` in this map, it would be more efficient to change it to a `ColSet`. Specifically, I think you can get rid of `projectedColsSet`, and change `projectedTupleCols` to an `opt.ColSet`.
___
*[pkg/sql/opt/norm/join.go, line 817 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36Sj-J8DPg-UCOLV6S:-M36Sj-J8DPg-UCOLV6T:bi1mdkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L817)):*
> ```Go
> 	}
> 	for _, f := range filters {
> 		// we're only interested in filters which reference a projected column
> ```

Comment should start with capital letter and end with a period.
___
*[pkg/sql/opt/norm/join.go, line 857 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36YQWJ88Rsyvsa9N-X:-M36YQWJ88Rsyvsa9N-Y:b-wc6n2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L857)):*
> ```Go
> 
> 	for _, p := range projections {
> 		if _, isTuple := p.Element.(*memo.TupleExpr); isTuple {
> ```

You might as well store the casted expression, since you'll just need to re-cast it later:
```
if tuple, isTuple := p.Element.(*memo.TupleExpr); isTuple {
  projectedCols[p.Col] = tuple
  ...
```
___
*[pkg/sql/opt/norm/join.go, line 864 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36WSd56SYQWhU8koZj:-M36WSd56SYQWhU8koZk:b-pmyjzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L864)):*
> ```Go
> 
> 	for _, f := range filters {
> 		if !f.ScalarProps().OuterCols.Intersection(projectedColsSet).Empty() {
> ```

You can use the function `Intersects` instead of checking whether the intersection is empty
___
*[pkg/sql/opt/norm/join.go, line 866 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36Wiu5-Aq3TxaVeQXj:-M36Wiu5-Aq3TxaVeQXk:bivjmnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L866)):*
> ```Go
> 		if !f.ScalarProps().OuterCols.Intersection(projectedColsSet).Empty() {
> 			eq, ok := f.Condition.(*memo.EqExpr)
> 			if ok {
> ```

[nit] can use `if eq, ok := f.Condition.(*memo.EqExpr); ok {`
___
*[pkg/sql/opt/norm/join.go, line 885 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36XCwG5DbIutJJ8iGh:-M36XCwG5DbIutJJ8iGi:b6508rr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L885)):*
> ```Go
> 				}
> 			} else {
> 				panic(errors.AssertionFailedf("cannot extract tuple equality from non-equality comparison"))
> ```

You probably want this panic to apply if any of the if statements above are false, not just the first one. This might be easier if you have a helper function that extracts the projected tuple and right tuple, and returns false if unable to do so.
___
*[pkg/sql/opt/norm/rules/join.opt, line 448 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZGzIDpsdCMfsur80:-M36ZGzIDpsdCMfsur81:bdm4nxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L448)):*
> ```
> # TODO(andyk): Add other join types.
> [HoistJoinProjectRight, Normalize]
> (InnerJoin | InnerJoinApply | LeftJoin | LeftJoinApply 
> ```

[nit] extra space
___
*[pkg/sql/opt/norm/rules/join.opt, line 469 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZPEr-wnTCYGlL0Ln:-M36ZPEr-wnTCYGlL0Lo:bdm4nxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L469)):*
> ```
> # input of the join.
> [HoistJoinProjectLeft, Normalize]
> (InnerJoin | InnerJoinApply | LeftJoin | LeftJoinApply 
> ```

[nit] extra space
___
*[pkg/sql/opt/norm/rules/join.opt, line 537 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZYcu0nKmfax6-pxJ:-M36ZYcu0nKmfax6-pxK:b-3gw8i0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L537)):*
> ```
> # ExtractProjectedTupleEqualityLeft detects joins which contain equalities between 
> # projected tuple variables and tuples, splits these tuples out into multiple 
> # equalities on the underlying columns, and hoists the project above the join this 
> ```

join this -> join. This
___
*[pkg/sql/opt/norm/rules/join.opt, line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZcUg6WdMcO9ff3D2:-M36ZcUg6WdMcO9ff3D3:bw3o4nz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L538)):*
> ```
> # projected tuple variables and tuples, splits these tuples out into multiple 
> # equalities on the underlying columns, and hoists the project above the join this 
> # can allow further simplification of project operators
> ```

End comment with a period.
___
*[pkg/sql/opt/norm/rules/join.opt, line 575 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36_J0r2iMJ6NKox1Xe:-M36_J0r2iMJ6NKox1Xf:bj4e7p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L575)):*
> ```
>                 $projections
>                 $on
>             ) 
> ```

[nit] try to make the formatting consistent between these two rules
___
*[pkg/sql/opt/norm/rules/join.opt, line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36_aPC8KKdK5zu47Cv:-M36_aPD27VlMNyo_ibm:bdi5wkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L643)):*
> ```
> )
> ```

[nit] extra lines
___
*[pkg/sql/opt/norm/testdata/rules/join, line 2914 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d6uVRC3ZQVIySkHfe:-M36aEF6EalERy-fFtWr:b-optumd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/testdata/rules/join#L2914)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Please look inside this file and follow the same format. Each rule gets its own `------` block.
You should also use `expect` or`expect-not` for the particular rule we're testing.

We should come up with tests that exercise just the left rule, and just the right rule, in addition to this. We also need tests for cases where the rule satisfies a subset of the requirements but does not fire (e.g. if a projected tuple column is used in multiple filters)
</blockquote></details>

Should also add some tests for the other types of joins such as inner and left join.


<!-- Sent from Reviewable.io -->

RoryBlevins committed:
opt:detect and extract projected tuple equalities

This commit adds a new rule to join normalisation that detects joins on tuple equalities between a tuple and a variable that is a projected tuple and splits them out into multiple equalities on the underlying columns, and hoists the project above the join. This allows us to simplify a number of queries where previously the projected tuple was preventing us merging projections.

Resolves: #43198

Release justification: low-risk performance improvement

RoryBlevins reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde, @RoryBlevins, and @rytaft)

___
*[pkg/sql/opt/norm/join.go, line 802 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36RwRI8qAGM7DD4tSN:-M8VtD92B1C4yOBenpUq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L802)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] comment should be wrapped at 80 characters.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/join.go, line 803 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36SATgDehMumxhetWX:-M8VtEM27tNVgvsvl4EL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L803)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This comment doesn't seem to match what's actually happening in this function.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/join.go, line 813 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36U8Qo30ZdY3SJyI2h:-M8VtEzIBGrBZpWnefoZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L813)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Since you're never actually using `p` in this map, it would be more efficient to change it to a `ColSet`. Specifically, I think you can get rid of `projectedColsSet`, and change `projectedTupleCols` to an `opt.ColSet`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/join.go, line 817 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36Sj-J8DPg-UCOLV6S:-M8VtGSy7D0IQ7CIySgX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L817)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Comment should start with capital letter and end with a period.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/join.go, line 857 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36YQWJ88Rsyvsa9N-X:-M8VtHOp2kZFX7TgqIes:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L857)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You might as well store the casted expression, since you'll just need to re-cast it later:
```
if tuple, isTuple := p.Element.(*memo.TupleExpr); isTuple {
  projectedCols[p.Col] = tuple
  ...
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/join.go, line 866 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36Wiu5-Aq3TxaVeQXj:-M8VtHxU2VNiJtiSAY5-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L866)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] can use `if eq, ok := f.Condition.(*memo.EqExpr); ok {`
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/join.go, line 885 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36XCwG5DbIutJJ8iGh:-M8VtIN530-flF6CZiZK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/join.go#L885)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

You probably want this panic to apply if any of the if statements above are false, not just the first one. This might be easier if you have a helper function that extracts the projected tuple and right tuple, and returns false if unable to do so.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/join.opt, line 448 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZGzIDpsdCMfsur80:-M8VtJ2a8P_1RabLyj-3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L448)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] extra space
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/join.opt, line 469 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZPEr-wnTCYGlL0Ln:-M8VtJWSBGU8j0AYa6mC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L469)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] extra space
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/join.opt, line 537 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZYcu0nKmfax6-pxJ:-M8VtKKP3sKqUPtnCINo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L537)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

join this -> join. This
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/join.opt, line 538 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36ZcUg6WdMcO9ff3D2:-M8VtKlEA44TyDyAqpcj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L538)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

End comment with a period.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/join.opt, line 575 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36_J0r2iMJ6NKox1Xe:-M8VtLWv8HAuoiLI1y7i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L575)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] try to make the formatting consistent between these two rules
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/rules/join.opt, line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M36_aPC8KKdK5zu47Cv:-M8VtLwz67IHmuxMj1eH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dcbe4839982898114d975f718ac3c0132f456c8e/pkg/sql/opt/norm/rules/join.opt#L643)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] extra lines
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde, @RoryBlevins, and @rytaft)

___
*[pkg/sql/opt/norm/join_funcs.go, line 598 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8WeS1H0g63ZIO4ISsj:-M8WeS1H0g63ZIO4ISsk:bq8rjv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/join_funcs.go#L598)):*
> ```Go
> 			projectedTupleCols.Add(p.Col)
> 		}
> 	}
> ```

I think we should add a fast path here: if `projectedTupleCols.Empty()`then return false. No reason to walk the filters if there are no projected tuples.
___
*[pkg/sql/opt/norm/join_funcs.go, line 599 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8WbvLl9bhQvVWaCj6E:-M8WbvLm1gfIl4fig477:bkuh46g) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/join_funcs.go#L599)):*
> ```Go
> 		}
> 	}
> 	for _, f := range filters {
> ```

This doesn't seem quite right to me. If there exists a filter that references a projected column that is not an "extractable tuple" then `CanExtractTupleEquality` will return true, and then the rule will host the `Project` outside the join, leaving an "orphaned" column reference. Also, we'll definitely need more tests to detect problems like this.
___
*[pkg/sql/opt/norm/join_funcs.go, line 616 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8Wd0bt1qQujll5Vzq2:-M8Wd0bt1qQujll5Vzq3:b4j8dam) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/join_funcs.go#L616)):*
> ```Go
> 	f *memo.FiltersItem, projectedCols opt.ColSet,
> ) bool {
> 	if f.Condition.Op() != opt.EqOp {
> ```

NIT: I'd structure it like this, since you're already doing casts anyway (no reason to make `Op` virtual calls):
```
eq, isEq := f.Condition.(*memo.EqExpr)
if !isEq || eq.Right.Op() != opt.TupleOp {
  return false
}

variable, isVar := eq.Left.(*memo.VariableExpr)
return isVar && projectedCols.Contains(variable.Col)
```
___
*[pkg/sql/opt/norm/join_funcs.go, line 664 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8WfOFF0DIh7AEiFo1i:-M8WfOFF0DIh7AEiFo1j:b-aij0vo) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/join_funcs.go#L664)):*
> ```Go
> func (c *CustomFuncs) rewriteTupleEqualityFilterItem(
> 	f memo.FiltersItem,
> 	newFilters memo.FiltersExpr,
> ```

I'm not sure why this is working. When you pass a slice in Go, you're just passing a "slice header", which is passed by value. Therefore, calling `append` below should mean that the new filter is not visible to the caller, because this function would only be modifying the local version of `newFilters`, not the caller's version. The usual Go pattern to do what you're trying to do would be to return the updated slice as a return value.
___
*[pkg/sql/opt/norm/rules/join.opt, line 528 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8WXaY4ExkgUkl2cTLO:-M8WXaY4ExkgUkl2cTLP:be5jz2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/rules/join.opt#L528)):*
> ```
> # This can allow further simplification of project operators.
> [ExtractProjectedTupleEqualityLeft, Normalize]
> (InnerJoin | LeftJoin | SemiJoin
> ```

The use case you're solving is important:
```
SELECT * FROM ab WHERE (a, b) IN (SELECT c, d FROM cd)
```
However, I think it'd be simpler and faster to match earlier, rather than waiting until after we've hosted the subquery. We should match on `Select` rather than joins (joins are kind of a pain b/c there are so many variants to think about, as well as the Left/Right variants), at this intermediate state:
```
select
 ├── columns: a:1!null b:2!null
 ├── scan ab
 │    └── columns: a:1!null b:2!null
 └── filters
      └── exists
           └── select
                ├── columns: column6:6
                ├── project
                │    ├── columns: column6:6
                │    ├── scan cd
                │    │    └── columns: c:3 d:4
                │    └── projections
                │         └── (c:3, d:4) [as=column6:6]
                └── filters
                     └── (a:1, b:2) = column6:6
```
You can use the `optsteps` command in a test file in order to see all the intermediate states. It's very helpful when building rules.
___
*[pkg/sql/opt/norm/rules/join.opt, line 533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8WVl7G8V2FYgwZsfDQ:-M8WVl7G8V2FYgwZsfDR:b-g3eu7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/rules/join.opt#L533)):*
> ```
>     $on:[
>         ...
>             (FiltersItem (Eq (Variable) (Tuple))) &
> ```

This match pattern is redundant with the code in `CanExtractTupleEquality`. It will create an unnecessary N**2 situation, where the matcher will loop over every filter condition, and then call `CanExtractTupleEquality`, which will then loop over every filter condition. I think here you only need:
```
$on:* & (CanExtractTupleEquality $projections $on)
```


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46186)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @RoryBlevins)

___
*[pkg/sql/opt/norm/join.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M2d4e8H0BvhkACgfWjN:-M8ldYws3L-IkASU4mwj:b-pwuush) ([raw file](https://github.com/cockroachdb/cockroach/blob/0847229ed13dc6da06e6f48fdb6d193737e2e838/pkg/sql/opt/norm/join.go#L832)):*
<details><summary><i>Previously, RoryBlevins wrote…</i></summary><blockquote>

I thought I had already run gofmt - but apparently not. Is there any documentation on how to run `crlfmt`?  I'm new to go and may be missing a convention, but the documentation seems to be a single line.
</blockquote></details>

I just updated the usage info at https://github.com/cockroachdb/crlfmt. Let me know if the command shown there doesn't work.

Depending on what editor you're using, you can also set it up to run automatically....
___
*[pkg/sql/opt/norm/join_funcs.go, line 658 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8le86k6L8JYSKZxWGN:-M8le86k6L8JYSKZxWGO:b-a3aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/join_funcs.go#L658)):*
> ```Go
> }
> 
> // rewriteTupleEqualityFilterItem akes a FilterItem that is an equality between
> ```

[nit] akes -> takes
[nit] FilterItem -> FiltersItem
___
*[pkg/sql/opt/norm/join_funcs.go, line 667 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/46186#-M8lLKQ5C-ZS4Bs2gr6A:-M8lLKQ5C-ZS4Bs2gr6B:bav6fmf) ([raw file](https://github.com/cockroachdb/cockroach/blob/622d77af0bfd4d45bd676ddb2514dea5f2738a00/pkg/sql/opt/norm/join_funcs.go#L667)):*
> ```Go
> 	projectedCols map[opt.ColumnID]*memo.TupleExpr, 
> 	) bool{
> 	if eq, ok := f.Condition.(*memo.EqExpr) ; ok {
> ```

[nit] extra space after cast before ";" (I think there are lint tests that catch this -- probably one of the failed tests in the CI build. You should probably configure your editor to run `crlfmt` on each save so you don't get all these linter errors -- see https://github.com/cockroachdb/crlfmt). 


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Closing in favor of https://github.com/cockroachdb/cockroach/pull/63780
